From 834c86d542391d698d5648929bf7a2f4ecf328a5 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Tue, 12 Aug 2008 10:32:24 -0400
Subject: [PATCH] lttng-instrumentation-ipc-tracepoints-probes

LTTng instrumentation ipc tracepoint probes

Create a module which declares ipc tracepoint probes, using markers.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
CC: Alexander Viro <viro@zeniv.linux.org.uk>
CC: 'Peter Zijlstra' <peterz@infradead.org>
CC: "Frank Ch. Eigler" <fche@redhat.com>
CC: 'Ingo Molnar' <mingo@elte.hu>
CC: 'Hideo AOKI' <haoki@redhat.com>
CC: Takashi Nishiie <t-nishiie@np.css.fujitsu.com>
CC: 'Steven Rostedt' <rostedt@goodmis.org>
CC: Masami Hiramatsu <mhiramat@redhat.com>
---
 ipc/Makefile    |    2 +-
 ipc/ipc-trace.c |   53 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 54 insertions(+), 1 deletions(-)
 create mode 100644 ipc/ipc-trace.c

diff --git a/ipc/Makefile b/ipc/Makefile
index 65c3843..eaed72e 100644
--- a/ipc/Makefile
+++ b/ipc/Makefile
@@ -8,4 +8,4 @@ obj-$(CONFIG_SYSVIPC_SYSCTL) += ipc_sysctl.o
 obj_mq-$(CONFIG_COMPAT) += compat_mq.o
 obj-$(CONFIG_POSIX_MQUEUE) += mqueue.o msgutil.o $(obj_mq-y)
 obj-$(CONFIG_IPC_NS) += namespace.o
-
+obj-$(CONFIG_TRACEPROBES) += ipc-trace.o
diff --git a/ipc/ipc-trace.c b/ipc/ipc-trace.c
new file mode 100644
index 0000000..4f3e7d7
--- /dev/null
+++ b/ipc/ipc-trace.c
@@ -0,0 +1,53 @@
+/*
+ * ipc/ipc-trace.c
+ *
+ * IPC tracepoint probes.
+ */
+
+#include <linux/module.h>
+#include <trace/ipc.h>
+
+static void probe_ipc_msg_create(long id, int flags)
+{
+	trace_mark(ipc_msg_create, "id %ld flags %d", id, flags);
+}
+
+static void probe_ipc_sem_create(long id, int flags)
+{
+	trace_mark(ipc_sem_create, "id %ld flags %d", id, flags);
+}
+
+static void probe_ipc_shm_create(long id, int flags)
+{
+	trace_mark(ipc_shm_create, "id %ld flags %d", id, flags);
+}
+
+int __init ipc_trace_init(void)
+{
+	int ret;
+
+	ret = register_trace_ipc_msg_create(probe_ipc_msg_create);
+	WARN_ON(ret);
+	ret = register_trace_ipc_sem_create(probe_ipc_sem_create);
+	WARN_ON(ret);
+	ret = register_trace_ipc_shm_create(probe_ipc_shm_create);
+	WARN_ON(ret);
+
+	return 0;
+}
+
+module_init(ipc_trace_init);
+
+void __exit ipc_trace_exit(void)
+{
+	unregister_trace_ipc_shm_create(probe_ipc_shm_create);
+	unregister_trace_ipc_sem_create(probe_ipc_sem_create);
+	unregister_trace_ipc_msg_create(probe_ipc_msg_create);
+	tracepoint_synchronize_unregister();
+}
+
+module_exit(ipc_trace_exit);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Mathieu Desnoyers");
+MODULE_DESCRIPTION("IPC Tracepoint Probes");
-- 
1.5.5.1

